### PR TITLE
Implement PIMPL for Xapian::Diversify

### PR DESCRIPTION
**Pull request by uppinder**
This PR implements the PIMPL idiom for Xapian::Diversify.

### CODE REVIEW

File: xapian-core/diversify/diversifyinternal.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#include <vector>
+
+/** Internal class for Diversify
+ */
+class Xapian::Diversify::Internal : public Xapian::Internal::intrusive_base {
+    /// Copies are not allowed
+    Internal(const Internal&);
+
+    /// Assignment is not allowed
+    void operator=(const Internal&);
```
[the remark this answers is not in the document]
Comment (uppinder) [reply xapian-core/diversify/diversifyinternal.h]: Thanks for pointing this out. Pushed a fixup commit for this.

File: xapian-core/diversify/diversifyinternal.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    /// Store docids of top k diversified documents
+    std::vector<Xapian::docid> main_dmset;
+
+  public:
+    /// Constructor for initialising diversification parameters
+    explicit Internal(unsigned int k_,
+		      double lambda_,
+		      double b_,
+		      double sigma_sqr_)
+    : k(k_), lambda(lambda_), b(b_), sigma_sqr(sigma_sqr_) {}
```
[the remark this answers is not in the document]
Comment (uppinder) [reply xapian-core/diversify/diversifyinternal.h]: Thanks. Modification pushed in a fixup commit.

File: xapian-core/diversify/diversifyinternal.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+     *
+     *  @param dmset	Set of points representing candidate diversifed
+     *			set of documents
+     */
+    double evaluate_dmset(const std::vector<Xapian::docid>& dmset);
+
+    /// Return diversified document set from given mset
+    Xapian::DocumentSet get_dmset(const MSet& mset);
+
+};
```
[the remark this answers is not in the document]
Comment (uppinder) [reply xapian-core/diversify/diversifyinternal.h]: Thanks for pointing this out, change pushed in a fixup commit.

File: xapian-core/include/xapian/diversify.h [hunk 4]
```diff
@@ -43,71 +43,37 @@ namespace Xapian {
...
+    Diversify& operator=(const Diversify& other);
 
-    /** Return difference of 'points' and current dmset
+    /** Move constructor.
      *
-     *  Return the difference of 'points' and the current diversified
-     *  document match set
-     *
-     *  @param dmset	Document set representing a diversified document set
+     * @param other	The object to move.
```
[the remark this answers is not in the document]
Comment (uppinder) [reply xapian-core/include/xapian/diversify.h]: Thanks, modification pushed in a fixup commit.

File: xapian-core/diversify/diversifyinternal.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  public:
+    /// Constructor for initialising diversification parameters
+    explicit Internal(unsigned int k_,
+		      double lambda_,
+		      double b_,
+		      double sigma_sqr_)
+    : k(k_), lambda(lambda_), b(b_), sigma_sqr(sigma_sqr_) {}
+
+    /// Destructor
+    ~Internal() {}
```
[the remark this answers is not in the document]
Comment (uppinder) [reply xapian-core/diversify/diversifyinternal.h]: Thanks for pointing this out. Remove the empty destructor in a fixup commit. Thanks for sharing the link as well.

File: xapian-core/include/xapian/diversify.h [hunk 6]
```diff
@@ -43,71 +43,37 @@ namespace Xapian {
...
      */
-    std::vector<Xapian::docid>
-    compute_diff_dmset(const std::vector<Xapian::docid>& dmset);
+    Diversify(Diversify&& other);
```
[the remark this answers is not in the document]
Comment (uppinder) [reply xapian-core/include/xapian/diversify.h]: Thanks for pointing this out. I've created definitions of copy ctor and move operator using ```= default```, which I think should suffice.

[the remark this answers is not in the document]
Comment (ojwb) [reply xapian-core/include/xapian/diversify.h]: Yes - the actual work is done by the smart pointer class which default should hook up.

**Comment by ojwb** [comment]
Thanks, merged.  The CI failures looks like they're just due to some connectivity problems.

**Pull request closed without merging**